### PR TITLE
Add sys logger.

### LINKED ISSUE #1406: implement sys logger for apisix

### PR DESCRIPTION
**Pull request by Akayeshmantha**
### Issues resolved

Fix #1406

### CODE REVIEW

**Comment by Akayeshmantha** [comment]
@membphis @moonming 

I think something wrong in trvis

Error: Failed installing dependency: https://luarocks.org/lua-resty-cookie-0.1.0-1.rockspec - Error fetching file: Failed downloading https://luarocks.org - Failed downloading https://luarocks.org/lua-resty-cookie-0.1.0-1.rockspec - /Users/travis/.cache/luarocks/https___luarocks.org/lua-resty-cookie-0.1.0-1.rockspec
make: *** [deps] Error 1
The command "$PWD/.travis/${OSNAME}_runner.sh do_install" failed and exited with 2 during .
Your build has been stopped.

But the module is available. Can we restart travis ?

**Comment by moonming** [comment]
I restart travis CI

**Comment by Akayeshmantha** [comment]
@membphis  @moonming 

It's  ready to revew.

**Comment by membphis** [comment]
By the way, the current implementation mechanism of resty-logger is not very suitable for APISIX.

In APISIX, we can create multiple routes and set different plug-in configurations, such as using different host and port for the plugin `syslog`.

The initialization of resty-logger is global, and it is impossible to use different resty-logger objects for different plugin configurations.

**Comment by Akayeshmantha** [comment]
so means we have to check another library than resty logger? @membphis

**Comment by membphis** [comment]
> so means we have to check another library than `resty logger`?

We can try to find other solutions, if there is no suitable, we need to modify the `resty-logger`.

**Comment by Akayeshmantha** [comment]
Yeah I think it will be ok if we have a new socket connection open like below ?

        location /m {
            log_by_lua '
                local logger_socket = require "resty.logger.socket"
                local logger = logger_socket:new()
                if not logger:initted() then
                    local ok, err = logger:init{
                        host = 'xxx',
                        port = 1234,
                        flush_limit = 1234,
                        drop_limit = 5678,
                    }
                    if not ok then
                        ngx.log(ngx.ERR, "failed to initialize the logger: ",
                                err)
                        return
                    end
                end

                local bytes, err = logger:log(msg)
                if err then
                    ngx.log(ngx.ERR, "failed to log message: ", err)
                    return
                end
            ';
        }`
`

**Comment by membphis** [comment]
We need a way to create or destroy logger objects, which requires complete lifecycle management.

only `create` is not enough.

**Comment by Akayeshmantha** [comment]
1. https://github.com/cloudflare/lua-resty-logger-socket/compare/master...whuben:master
2. https://github.com/bsc-s2/lua-resty-logger-socket/commit/d5c2bf200bf6295655f6db33a36ee9b71e705cec

from these 2 what is the one we are going to pick when moving forward 
@membphis  @moonming

**Comment by membphis** [comment]
The two implementation methods are similar, but the second one has test cases, so I prefer it.

https://github.com/bsc-s2/lua-resty-logger-socket/commit/d5c2bf200bf6295655f6db33a36ee9b71e705cec

**Comment by Akayeshmantha** [comment]
Sure @membphis

**Comment by Akayeshmantha** [comment]
@membphis updated with the latest library

File: apisix/plugins/syslog.lua [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    name = plugin_name,
+    schema = schema,
+}
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+function _M.flush_syslog(logger)
+        local ok, err = logger:flush(logger)
```
Comment (membphis) [inline apisix/plugins/syslog.lua:52]: bad indentation: 4 spaces

File: apisix/plugins/syslog.lua [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+function _M.log(conf)
+    local entry = log_util.get_full_log(ngx)
+
+    if not entry.route_id then
+        core.log.error("failed to obtain the route id for sys logger")
+        return
+    end
+
+    local logger, err = logger_socket:new({
```
Comment (membphis) [inline apisix/plugins/syslog.lua:66]: that is not a good way. it is expensive to create a logger object for each request.

we can use lrucache here, you can take a look at this comment:

https://github.com/api7/lua-resty-logger-socket/pull/3#issuecomment-612538990

Comment (Akayeshmantha) [reply apisix/plugins/syslog.lua]: updated and thanks alot

File: apisix/plugins/syslog.lua [hunk 3]
```diff
@@ -63,23 +68,26 @@ function _M.log(conf)
...
-        drop_limit = conf.drop_limit,
-        timeout = conf.timeout,
-        sock_type = conf.sock_type,
-        max_retry_times = conf.max_retry_times,
-        retry_interval = conf.retry_interval,
-        pool_size = conf.pool_size,
-        tls = conf.tls,
-    })
+    -- fetch it from lrucache
+    local logger, err =  lrucache("127.0.0.1", "29999",
```
Comment (membphis) [inline apisix/plugins/syslog.lua:72]: We can use `api_ctx.conf_type .. "#" .. api_ctx.conf_id` as key, `api_ctx.conf_version` as version for this lrucache use case.

`local logger, err =  lrucache("127.0.0.1", "29999",`
to
`local logger, err =  lrucache(api_ctx.conf_type .. "#" .. api_ctx.conf_id, api_ctx.conf_version, ...`

Here is an example: https://github.com/apache/incubator-apisix/blob/master/apisix/core/lrucache.lua#L199

Comment (Akayeshmantha) [reply apisix/plugins/syslog.lua]: fix.

**Comment by Akayeshmantha** [comment]
@membphis  resolve the conflicts

**Comment by Akayeshmantha** [comment]
@membphis can we merge this 🙂

**Comment by Akayeshmantha** [comment]
Hi guys @moonming  @membphis  can we merge this so I can work on this issue and finish things on sys logger.
https://github.com/apache/incubator-apisix/issues/1500

**Comment by moonming** [comment]
> Hi guys @moonming @membphis can we merge this so I can work on this issue and finish things on sys logger.
> #1500

lgtm

**Pull request merged**